File: app/src/main/java/com/example/sunkai/heritage/Activity/RegistActivity.java
package com.example.sunkai.heritage.Activity;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.example.sunkai.heritage.ConnectWebService.HandleUser;
import com.example.sunkai.heritage.R;

import java.util.HashMap;

public class RegistActivity extends AppCompatActivity implements View.OnClickListener,View.OnFocusChangeListener {

    private TextView regist_actitivy_username_textView;
    private EditText regist_actitivy_username_editText;
    private TextView regist_actitivy_password_textView;
    private EditText regist_actitivy_password_editText;
    private TextView regist_actitivy_insure_textView;
    private EditText regist_actitivy_insure_editText;
    private TextView regist_actitivy_question_textView;
    private EditText regist_actitivy_question_editText;
    private TextView regist_actitivy_answer_textView;
    private EditText regist_actitivy_answer_editText;
    private Button regist_activity_regist_button;
    private ProgressBar progressBar;
    private ActionBar actionBack;
    private HashMap<EditText,TextView> editTextToTextViewHashMap;

    String userName, userPassword, findPasswordQuestion, findPasswordAnswer;
    private Button regist_activity_cancel_button;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_regist);
        initView();
    }

    private void initView() {
        editTextToTextViewHashMap=new HashMap<>();
        regist_actitivy_username_textView = (TextView) findViewById(R.id.regist_actitivy_username_textView);
        regist_actitivy_username_editText = (EditText) findViewById(R.id.regist_actitivy_username_editText);
        regist_actitivy_username_editText.setOnFocusChangeListener(this);
        editTextToTextViewHashMap.put(regist_actitivy_username_editText,regist_actitivy_username_textView);
        regist_actitivy_password_textView = (TextView) findViewById(R.id.regist_actitivy_password_textView);
        regist_actitivy_password_editText = (EditText) findViewById(R.id.regist_actitivy_password_editText);
        editTextToTextViewHashMap.put(regist_actitivy_password_editText,regist_actitivy_password_textView);
        regist_actitivy_password_editText.setOnFocusChangeListener(this);
        regist_actitivy_insure_textView = (TextView) findViewById(R.id.regist_actitivy_insure_textView);
        regist_actitivy_insure_editText = (EditText) findViewById(R.id.regist_actitivy_insure_editText);
        editTextToTextViewHashMap.put(regist_actitivy_insure_editText,regist_actitivy_insure_textView);
        regist_actitivy_insure_editText.setOnFocusChangeListener(this);
        regist_actitivy_question_textView = (TextView) findViewById(R.id.regist_actitivy_question_textView);
        regist_actitivy_question_editText = (EditText) findViewById(R.id.regist_actitivy_question_editText);
        editTextToTextViewHashMap.put(regist_actitivy_question_editText,regist_actitivy_question_textView);
        regist_actitivy_question_editText.setOnFocusChangeListener(this);
        regist_actitivy_answer_textView = (TextView) findViewById(R.id.regist_actitivy_answer_textView);
        regist_actitivy_answer_editText = (EditText) findViewById(R.id.regist_actitivy_answer_editText);
        editTextToTextViewHashMap.put(regist_actitivy_answer_editText,regist_actitivy_answer_textView);
        regist_actitivy_answer_editText.setOnFocusChangeListener(this);
        regist_activity_regist_button = (Button) findViewById(R.id.activity_regist_regist_button);
        regist_activity_regist_button.setOnClickListener(this);
        progressBar = (ProgressBar) findViewById(R.id.progressBar);
        progressBar.setOnClickListener(this);
        regist_activity_cancel_button = (Button) findViewById(R.id.regist_activity_cancel_button);
        regist_activity_cancel_button.setOnClickListener(this);
        actionBack=getSupportActionBar();
        actionBack.setDisplayHomeAsUpEnabled(true);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.activity_regist_regist_button:
                submit();
                break;
            case R.id.regist_activity_cancel_button:
                this.setResult(0);
                finish();
            default:
                break;
        }
    }

    @Override
    public void onFocusChange(View v, boolean hasFocus) {
        switch (v.getId()){
            case R.id.regist_actitivy_username_editText:
            case R.id.regist_actitivy_password_editText:
            case R.id.regist_actitivy_insure_editText:
            case R.id.regist_actitivy_question_editText:
            case R.id.regist_actitivy_answer_editText:
                changeTextviewStatus((EditText)v,hasFocus);
                break;
            default:
                    break;
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void changeTextviewStatus(EditText editText,Boolean hasFocus){
        TextView textView=editTextToTextViewHashMap.get(editText);
        if(textView==null)
            return;
        if(hasFocus)
            textView.setTextColor(getResources().getColor(R.color.colorPrimary));
        else
            textView.setTextColor(getResources().getColor(R.color.deepGrey));
    }
    private void submit() {
        // validate
        String editText = regist_actitivy_username_editText.getText().toString().trim();
        if (TextUtils.isEmpty(editText)) {
            regist_actitivy_username_editText.setError("用户名不能为空");
            regist_actitivy_username_editText.requestFocus();
            return;
        }

        editText = regist_actitivy_password_editText.getText().toString().trim();
        if (TextUtils.isEmpty(editText)) {
            regist_actitivy_password_editText.setError("密码不能为空");
            regist_actitivy_password_editText.requestFocus();
            return;
        }

        editText = regist_actitivy_insure_editText.getText().toString().trim();
        if (TextUtils.isEmpty(editText)) {
            regist_actitivy_insure_editText.setError("确认密码不能为空");
            regist_actitivy_insure_editText.requestFocus();
            return;
        }

        editText = regist_actitivy_question_editText.getText().toString().trim();
        if (TextUtils.isEmpty(editText)) {
            regist_actitivy_question_editText.setError("密码召回问题不能为空");
            regist_actitivy_question_editText.requestFocus();
            return;
        }

        editText = regist_actitivy_answer_editText.getText().toString().trim();
        if (TextUtils.isEmpty(editText)) {
            regist_actitivy_answer_editText.setError("密码找回答案不能为空");
            regist_actitivy_answer_editText.requestFocus();
            return;
        }
        String password=regist_actitivy_password_editText.getText().toString();
        String insurePassword=regist_actitivy_insure_editText.getText().toString();
        if(!(password.equals(insurePassword))){
            Toast.makeText(this,"密码输入不一致",Toast.LENGTH_SHORT).show();
            regist_actitivy_insure_editText.requestFocus();
            return;
        }

        // TODO validate success, do something
        userName = regist_actitivy_username_editText.getText().toString().trim();
        userPassword = regist_actitivy_password_editText.getText().toString().trim();
        findPasswordQuestion = regist_actitivy_question_editText.getText().toString().trim();
        findPasswordAnswer = regist_actitivy_answer_editText.getText().toString().trim();
        progressBar.setVisibility(View.VISIBLE);
        regist_actitivy_username_editText.setEnabled(false);
        regist_actitivy_password_editText.setEnabled(false);
        regist_actitivy_insure_editText.setEnabled(false);
        regist_actitivy_question_editText.setEnabled(false);
        regist_actitivy_answer_editText.setEnabled(false);
        regist_activity_cancel_button.setVisibility(View.GONE);
        regist_activity_regist_button.setEnabled(false);
        new Thread(userRegist).start();
    }


    Runnable userRegist = new Runnable() {
        @Override
        public void run() {
            try {
                Thread.sleep(800);
            }
            catch (InterruptedException e){
                e.printStackTrace();
            }
            int result= HandleUser.User_Regist(userName,userPassword,findPasswordQuestion,findPasswordAnswer);
            Message msg=new Message();
            msg.what=result;
            final Handler userRegistHandler=new Handler(getMainLooper()){
                @Override
                public void handleMessage(Message msg){
                    if(msg.what==1){
                        progressBar.setVisibility(View.VISIBLE);
                        regist_activity_cancel_button.setVisibility(View.VISIBLE);
                        Intent intent=new Intent();
                        intent.putExtra("userName",userName);
                        intent.putExtra("passWord",userPassword);
                        RegistActivity.this.setResult(1,intent);
                        Toast.makeText(RegistActivity.this,"注册成功",Toast.LENGTH_SHORT).show();
                        finish();
                    }
                    else if(msg.what==0){
                        Toast.makeText(RegistActivity.this,"已有该用户",Toast.LENGTH_SHORT).show();
                    }
                    else{
                        Toast.makeText(RegistActivity.this,"注册失败",Toast.LENGTH_SHORT).show();
                    }
                    progressBar.setVisibility(View.GONE);
                    regist_actitivy_username_editText.setEnabled(true);
                    regist_actitivy_password_editText.setEnabled(true);
                    regist_actitivy_insure_editText.setEnabled(true);
                    regist_actitivy_question_editText.setEnabled(true);
                    regist_actitivy_answer_editText.setEnabled(true);
                    regist_activity_cancel_button.setVisibility(View.VISIBLE);
                    regist_activity_regist_button.setEnabled(true);
                }
            };
            userRegistHandler.sendMessage(msg);
        }
    };


}

File: app/src/main/java/com/example/sunkai/heritage/Activity/RegistActivity.kt
package com.example.sunkai.heritage.activity

import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.transition.Slide
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.core.view.children
import com.example.sunkai.heritage.activity.baseActivity.BaseTakeCameraActivity
import com.example.sunkai.heritage.connectWebService.HandleUser
import com.example.sunkai.heritage.R
import com.example.sunkai.heritage.tools.MakeToast.toast
import com.example.sunkai.heritage.tools.WindowHelper
import com.example.sunkai.heritage.tools.encryptionPassWord
import com.example.sunkai.heritage.tools.toByteArray
import com.example.sunkai.heritage.value.ERROR
import com.example.sunkai.heritage.value.PASSWORD
import com.example.sunkai.heritage.value.USER_NAME
import kotlinx.android.synthetic.main.activity_regist.*
import java.util.*


class RegistActivity : BaseTakeCameraActivity(), View.OnClickListener, TextWatcher {

    private var userImageBitmap: Bitmap? = null

    private var isUploadImage = false

    private val views: ArrayList<View>by lazy {
        ArrayList<View>()
    }
    private val errorMessage: Array<String> by lazy {
        arrayOf("用户名不能为空",
                "密码不能为空",
                "确认密码不能为空",
                "密码召回问题不能为空",
                "密码找回答案不能为空",
                "密码输入不一致")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setBackgroundDrawable(null)
        startAnimation()
        setContentView(R.layout.activity_regist)
        initView()
    }

    override fun onStart() {
        super.onStart()
        setWindowFullScreen()
    }

    private fun startAnimation() {
        window.enterTransition = Slide(Gravity.END)
    }

    private fun setWindowFullScreen() {
       WindowHelper.setWindowFullScreen(this)
    }

    private fun initView() {
        addAllViews(registAllViewLinearLatyout)
        setBackGround()
        setAllViewsOnclick()
        regist_actitivy_password_editText.addTextChangedListener(this)
        regist_actitivy_insure_editText.addTextChangedListener(this)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    //将所有的view放入到list
    private fun addAllViews(viewgroup: ViewGroup) {
        viewgroup.children.toList().forEach { if (it is ViewGroup) addAllViews(it) else views.add(it) }
    }

    private fun setBackGround() {
        val calendar = Calendar.getInstance()
        val hour = calendar.get(Calendar.HOUR_OF_DAY)
        if (hour >= 18 || hour <= 9) {
            activityRegistBackground.setBackgroundResource(R.mipmap.at_night_background)
        } else {
            activityRegistBackground.setBackgroundResource(R.mipmap.day_background)
        }
    }

    override fun setImageToImageView(bitmap: Bitmap) {
        glide.load(bitmap).into(registUserImage)
        userImageBitmap = bitmap
        isUploadImage = true

    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.activity_regist_regist_button -> submit()
            R.id.registCancel -> {
                this.setResult(CANCEL)
                onBackPressed()
            }
        }
    }

    override fun getNeedOpenChooseImageView(): Array<View> {
        return arrayOf(registUserImage)
    }


    private fun setViewsIsEnable(isEnable: Boolean) {
        views.forEach {
            it.isEnabled = isEnable
        }
    }

    private fun setViewsIsEnable() {
        setViewsIsEnable(true)
    }

    private fun setViewsIsDisable() {
        setViewsIsEnable(false)
    }

    private fun judgeViewsTextIsEmpty(): Boolean {
        var OK = false
        var count = 0
        for (view in views) {
            //遍历所有的edittext，发现为空的时候显示对应的提示
            if (view is EditText) {
                val text = view.text.toString().trim()
                if (TextUtils.isEmpty(text)) {
                    view.error = errorMessage[count]
                    OK = true
                }
                count++
            }
        }
        return OK
    }

    private fun judgePasswordIsSame(): Boolean {
        val password = regist_actitivy_password_editText.text.toString()
        val insurePassword = regist_actitivy_insure_editText.text.toString()
        return if (password != insurePassword) {
            regist_actitivy_password_editText.error = "密码输入不一致"
            regist_actitivy_insure_editText.error = "密码输入不一致"
            true
        } else {
            regist_actitivy_password_editText.error = null
            regist_actitivy_insure_editText.error = null
            false
        }
    }

    private fun setAllViewsOnclick() {
        views.forEach { view -> view.setOnClickListener(this) }
    }

    private fun submit() {
        setViewsIsDisable()
        if (judgeViewsTextIsEmpty() || judgePasswordIsSame()) {
            setViewsIsEnable()
            return
        }
        //开始注册
        val userName = regist_actitivy_username_editText.text.toString().trim()
        val userPassword = regist_actitivy_password_editText.text.toString().trim()
        val findPasswordQuestion = regist_actitivy_question_editText.text.toString().trim()
        val findPasswordAnswer = regist_actitivy_answer_editText.text.toString().trim()
        requestHttp {
            userRegist(userName, userPassword, findPasswordQuestion, findPasswordAnswer)
        }
    }

    private fun userRegist(userName: String, userPassword: String, findPasswordQuestion: String, findPasswordAnswer: String) {
        val userPasswordDecript = (infoToRSA(userPassword) ?: return)
        val findPasswordAnswerDecript = (infoToRSA(findPasswordAnswer) ?: return)
        val result = if (isUploadImage) {
            val bitmap = userImageBitmap ?: return
            HandleUser.User_Regist(userName, userPasswordDecript, findPasswordQuestion, findPasswordAnswerDecript, bitmap.toByteArray())
        } else {
            HandleUser.User_Regist(userName, userPasswordDecript, findPasswordQuestion, findPasswordAnswerDecript, null)
        }
        runOnUiThread {
            when (result) {
                SUCCESS -> {
                    val intent = Intent()
                    intent.putExtra(USER_NAME, userName)
                    intent.putExtra(PASSWORD, userPassword)
                    setResult(SUCCESS, intent)
                    toast("注册成功")
                    finish()
                }
                HAD_USER -> toast("已有该用户")
                else -> toast("注册失败")
            }
            setViewsIsEnable()
        }
    }

    private fun infoToRSA(infos: String): String? {
        val encrtData = encryptionPassWord(infos)
        return if (ERROR == encrtData) {
            null
        } else encrtData
    }

    override fun afterTextChanged(s: Editable?) {
        judgePasswordIsSame()
    }

    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
    }

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
    }

    companion object {
        const val CANCEL = 0
        const val SUCCESS = 1
        const val HAD_USER = 0
    }
}

